### PR TITLE
Allow ID attributes on any elements

### PR DESCRIPTION
**Pull request by pento**
## Description

The ID attribute is frequently used to create intra-page links, or to link to specific parts of a page.

Removing them when we convert existing posts to blocks causes those links to suddenly start failing.

It's worth noting that this PR intentionally _does not_ create UI for editing the `id` attribute. The workflow for adding an `id` attribute in Classic involves switching to Text view, the workflow for Gutenberg currently remains the same. This PR is solely to address the data loss that can occur when a post is converted.

## How has this been tested?

Create a post in the Classic Editor with this content:

`<h2 id="foo">Bar</h2>`

Open that post in Gutenberg, and convert it to blocks.

Check the Code view, to see that the `id` attribute is still set correctly.

Preview the post, and add the `#foo` anchor to the URL, and check that the page jumps to the heading.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Review by gziolo — APPROVED** [review]
This change looks good and I agree with the reasoning behind it.

I would like to make sure @aduth or @iseulde are okey with this change.

**Review by youknowriad — COMMENTED** [review]
We should definitely avoid dropping the `id` for blocks supporting `anchor` like the heading but it's pointless to keep it for the other blocks because It will be dropped anyway by the serialization. An alternative would be to support `anchor` for blocks by default (like we do for custom classNames) but I think we decided against when we originally introduced the anchor. The argument is, the anchor doesn't make much sense for most blocks and is more important for headings (false by default).

**Comment by aduth** [comment]
> it's pointless to keep it for the other blocks because It will be dropped anyway by the serialization.

This is an important note, and maybe also concerning that our tests don't verify that serialization won't cause the attributes to be dropped anyways.

**Comment by pento** [comment]
Thanks for the feedback! I changed it only allow the `id` attribute on blocks that support `anchor`.

File: packages/blocks/src/api/raw-handling/test/utils.js [hunk 1]
```diff
@@ -109,6 +109,12 @@ describe( 'removeInvalidHTML', () => {
 		expect( removeInvalidHTML( input, schema ) ).toBe( output );
 	} );
 
+	it( 'should remove id attributes', () => {
+		const input = '<p id="foo">test</p>';
+		const output = '<p>test</p>';
+		expect( removeInvalidHTML( input, schema ) ).toBe( output );
+	} );
```
Comment (youknowriad) [inline packages/blocks/src/api/raw-handling/test/utils.js]: Is it possible to add a unit tests for headings (where we keep the attribute?)

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by youknowriad** [comment]
One question that crosses my mind here is that the `anchor` support is added by a hook, so I was wondering if it's possible to extend the raw handler by a plugin as well to add this logic of keeping the `id`? Or should we add a filter/hook to the raw handler to allow this kind of tweaks?

File: packages/blocks/src/api/raw-handling/test/utils.js (hunk 1 above)
Comment (pento) [reply packages/blocks/src/api/raw-handling/test/utils.js]: Not without moving things around, or copy/pasting code. Testing that functionality requires access to `getRawTransformations()`, which currently isn't exported.

**Comment by pento** [comment]
If a plugin's block declares that it supports `anchor`, then the HTML that its raw transform generates should get to keep the `id` attribute.

There's probably going to be a need to make this [pluggable](https://giphy.com/gifs/dog-nirvana-doggie-dTJd5ygpxkzWo), particularly for sites that generate weird HTML that they want to keep.

**Comment by noisysocks** [comment]
Fixes #7335.

**Pull request merged**